### PR TITLE
GCW-3663-Add offer response table

### PR DESCRIPTION
**Pull request by shreyas098**
### Ticket Link: 

https://jira.crossroads.org.hk/browse/GCW-3663

### What does this PR do?

FEATURE: Adds offer response table

### CODE REVIEW

File: app/controllers/api/v1/offer_responses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        param :offer_response, Hash, required: true do
+          param :user_id, String, desc: "Id of user"
+          param :offer_id, String, desc: "Id of offer"
+        end
+      end
+
+      api :POST, "/v1/offer_responses","Create a offer Response"
+      def create
+        user_id, offer_id = params["offer_response"].values_at('user_id', 'offer_id')
+        return  render json: { errors: "Offer and User Id required"}, status: 422 unless user_id.present? && offer_id.present?
```
Comment (steveyken) [inline app/controllers/api/v1/offer_responses_controller.rb:25]: Model validation should pick this up already so not required

File: app/controllers/api/v1/offer_responses_controller.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        @offer_response=OfferResponse.new
+        @offer_response.user=User.find_by_id(user_id)
+        @offer_response.offer=Offer.find_by_id(offer_id)
```
Comment (steveyken) [inline app/controllers/api/v1/offer_responses_controller.rb:29]: rails active-record should handle the user assignment, so this isn't needed

**Review by steveyken — CHANGES_REQUESTED** [review]
I haven't run through everything but made a few comments... let's chat on Slack about this so far

File: app/controllers/api/v1/ability.rb [hunk 3]
```diff
@@ -96,7 +97,11 @@ def beneficiary_abilities
       end
 
       def canned_response_abilities
-        can %I[index create update destroy show], CannedResponse if can_manage_canned_response?
+        can %i[index create update destroy show], CannedResponse if can_manage_canned_response?
+      end
+
+      def offer_response_abilities
+        can [:create, :index], OfferResponse
```
[the remark this answers is not in the document]
Comment (patrixr) [reply app/controllers/api/v1/ability.rb:104]: @steveyken should we maybe only allow creation on offers that have been shared ? Otherwise I could start inputting random IDs and get access to them

File: db/migrate/20210620222418_create_offer_responses.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  def change
+    create_table :offer_responses do |t|
+      t.integer   :user_id
+      t.integer   :offer_id
+
+      t.timestamps
+    end
+
+  add_foreign_key :offer_responses, :users, column: :user_id
+  add_foreign_key :offer_responses, :offers, column: :offer_id
```
Comment (steveyken) [inline db/migrate/20210620222418_create_offer_responses.rb:11]: would be good to add 2 indexes for offer_id and user_id also

File: app/controllers/api/v1/ability.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (steveyken) [reply app/controllers/api/v1/ability.rb:104]: Good call... @shreyas098  will include a block to call `Shared.shareable_resource?` as for `messages_abilities` below

File: app/models/concerns/message_subscriptions.rb [hunk 5]
```diff
@@ -47,7 +47,11 @@ def required_staff_permission(klass)
     if ['Offer', 'Item'].include?(klass)
       return 'can_manage_offer_messages'
     end
-    
+
+    if ['OfferResponse'].include?(klass)
+      return 'can_manage_offer_messages'
+    end
+
```
Comment (patrixr) [inline app/models/concerns/message_subscriptions.rb:54]:
```suggestion
    if ['Offer', 'Item', 'OfferResponse'].include?(klass)
      return 'can_manage_offer_messages'
    end
```

File: app/controllers/api/v1/ability.rb [hunk 6]
```diff
@@ -96,7 +97,12 @@ def beneficiary_abilities
       end
 
       def canned_response_abilities
-        can %I[index create update destroy show], CannedResponse if can_manage_canned_response?
+        can %i[index create update destroy show], CannedResponse if can_manage_canned_response?
+      end
+
+      def offer_response_abilities
+        can [:create, :index], OfferResponse, user_id: @user.id
```
Comment (patrixr) [inline app/controllers/api/v1/ability.rb:104]: This permission rule would still allow me to create an `offer_response` record to an offer that isn't shared

Comment (patrixr) [reply app/controllers/api/v1/ability.rb:104]: (do write a test for that)

**Review by patrixr — CHANGES_REQUESTED** [review]
@shreyas098 Things looking good overall, just added a couple of comments to be looked at. I recommend you have Steve take another look as I'm afraid I might miss some things

File: app/controllers/api/v1/ability.rb (hunk 6 above)
Comment (shreyas098) [reply app/controllers/api/v1/ability.rb:104]: sure , working on it

File: app/controllers/api/v1/offer_responses_controller.rb [hunk 7]
```diff
@@ -19,27 +19,22 @@ class  OfferResponsesController < Api::V1::ApiController
         end
       end
 
-      api :POST, "/v1/offer_responses","Create a offer Response"
+      api :POST, "/v1/offer_responses", "Create a Offer Response"
       def create
-        user_id, offer_id = params["offer_response"].values_at('user_id', 'offer_id')
-        return  render json: { errors: "Offer and User Id required"}, status: 422 unless user_id.present? && offer_id.present?
+        record = OfferResponse.find_or_create_by(offer_responses_params)
```
Comment (steveyken) [inline app/controllers/api/v1/offer_responses_controller.rb:24]: `CanCanCan` already gives you `@offer_response` with the params loaded, you don't need to do a `find_or_create_by` here. Just do `if @offer_response.save`

File: app/models/concerns/message_subscriptions.rb [hunk 8]
```diff
@@ -47,7 +47,11 @@ def required_staff_permission(klass)
     if ['Offer', 'Item'].include?(klass)
       return 'can_manage_offer_messages'
     end
-    
+
+    if ['OfferResponse'].include?(klass)
```
Comment (steveyken) [inline app/models/concerns/message_subscriptions.rb:51]: Why not make the permission `can_manage_offer_response_messages` and then you don't need this extra `if` statement, you can rely on the convention below to find it

File: app/controllers/api/v1/ability.rb (hunk 6 above)
Comment (steveyken) [reply app/controllers/api/v1/ability.rb:104]: Yes, `user_id` is correct here but we need to scope `offer_id` also. It should be the list of offers that the user is has created.

```
can [:create, :index], OfferResponse, { user_id: @user_id, offer: { created_by_id: @user.id } }
```

**Review by steveyken — COMMENTED** [review]
Looking much better but a few more things to tidy up.

File: app/controllers/api/v1/ability.rb (hunk 6 above)
Comment (patrixr) [reply app/controllers/api/v1/ability.rb:104]: @steveyken I think you're thinking of a different rule here. I think here `offer_id` is scoped to the shared offers

i.e

```ruby
can [:create, :index], OfferResponse, { user_id: @user_id } do |offer_response|
  Shareable.non_expired.find_by(
    resource_type: 'Offer',
    resource_id: offer_response.offer_id
  ).present?
end
```

**Comment by steveyken** [comment]
Ah yes, that's right. Good to scope to shared offers only

File: app/controllers/api/v1/offer_responses_controller.rb (hunk 7 above)
Comment (patrixr) [reply app/controllers/api/v1/offer_responses_controller.rb:24]: + 1

File: app/controllers/api/v1/ability.rb [hunk 9]
```diff
@@ -96,7 +97,21 @@ def beneficiary_abilities
...
+      def offer_response_abilities
+        can :create, OfferResponse do |offer_response|
+          offer_response.user_id == @user_id &&
+            Shareable.non_expired.find_by(
+              resource_type: 'Offer',
+              resource_id: offer_response.offer_id
+            ).present?
+        end
+
+        can %i[index show], OfferResponse
```
Comment (patrixr) [inline app/controllers/api/v1/ability.rb:112]:
```suggestion
        can %i[index show], OfferResponse, { user_id: @user_id }
```
A user should only be allowed to view his/her records

Comment (patrixr) [reply app/controllers/api/v1/ability.rb:112]: Note: make sure there's a test for it : )

File: db/migrate/20210620222418_create_offer_responses.rb (hunk 4 above)
Comment (patrixr) [reply db/migrate/20210620222418_create_offer_responses.rb:11]: @shreyas098 we probably need a unique composite index to avoid duplicates. And we need tests for duplicates generally

**Review by patrixr — CHANGES_REQUESTED** [review]

**Review by patrixr — APPROVED** [review]
Looks good @shreyas098 ! Thanks for all the work you've done on this !

**Review by swatijadhav — APPROVED** [review]

**Comment by steveyken** [comment]
This looks good now. Thanks @shreyas098
